### DISCUSSION TITLE
Cached `checksumAddress`

### DISCUSSION THREAD

**Discussion by kyscott18** [wevm/viem#2652] [Idea / Feature Request] [answered]
`checksumAddress` is a relatively expensive operation because it has to perform the `keccak256` hash function on all input. 

One option to speed it up would be to add an LRU cache that can potentially skip the expensive hash function when it gets a "hit". [This implementation](https://github.com/ponder-sh/ponder/blob/kjs/checksum-cache/packages/core/src/utils/checksumAddress.ts) led to a 77x performance improvement.

`checksumAddress` is also internally used in a few other places, namely `decodeAbiParameters`. Would you be open to a contribution to add this to viem?

**Comment by jxom** [comment, marked as answer]
`checksumAddress` already has an LRU cache.

https://github.com/wevm/viem/blob/main/src/utils/address/getAddress.ts#L34

How does this improve performance by 77x?

Update: I just realised that our LRU implementation doesn’t keep track of the keys touched to flush. Can update to fix.

**Reply by kyscott18** [reply]
Ah I actually didn't know that you had implemented this already, always one step ahead lol. I was benchmarking against v1.16 which didn't have the LRU cache. Thanks!
